### ISSUE TITLE
Remap Node exceeds its bounds

### ISSUE THREAD

**Issue by ProbablePrime** [Frooxius/NeosPublic#945]
Hey all, 

Spoke to a few people in Neos about this one. Some people I spoke to said that Coffee or Froox might already be aware but I wanted to log the GH issue for transparency and tracking purposes.

## Reproduction Case

1. Spawn a remap node
1. Enter 1 into the In Min Input
1. Enter 10 into the In Max Input
1. Enter 0 into the Out Min Input
1. Enter 1 into the Out Max Input
1. Enter 11 into the value input
1. Observe that the output exceeds the value of Out Max.

## The problems this causes

To me this seems like a bug, Remap's whole job is to constrain a number that's inputted between Out Min and Out Max. If it's exceeding Out Max. It just seems like a bug.

## Possible Solutions
- Internally clamp the output of Remap between OutMin and OutMax and clamp always
- Add a new input to the remap node called "Clamp" which is a boolean and enables this clamping selectively
- Make a new "Clamp Remap" node which acts like the first solution.

**Comment by mralext20** [comment]
my thoughts on this is that a new clamp bool would be the best solution, as this functionality is useful in some places.

**Comment by H3BO3** [comment]
Remap functions are designed to go out of bounds, it's up to the coder to clamp the input or output values if necessary - It's not a bug, and always clamping would break many existing and future applications. Interested to see what Froox says about an innate clamp boolean though.

**Comment by sauermon** [comment]
I believe an extra bool input for clamped output would be quite handy in my opinion, since currently if I want to have a clamped remap output, I have to use a clamp none either before or after the remap.

**Comment by Frooxius** [comment]
This is not a bug. Remap node is essentially fancy remapping of linear interpolation, which means that you can extrapolate values as well if you go outside of the range (similarly to an Unclamped Lerp).

There's a planned solution, which will add "Unclamped Remap" that'll function as the current one and convert all existing instances to it (otherwise we risk breaking stuff that uses the extrapolation functionality) and then add new "Remap" which has the clamping. Adding "Clamp Remap" would be inconsistent with Lerps.

**Comment by H3BO3** [comment]
@Frooxius Wouldn't it be better to have Remap stay the same, and have "Clamped Remap" be the new node? Programmers may go for the one named Remap and expect the same functionality as they're used to, not to mention the positive naming would be more intuitive.

**Comment by Frooxius** [comment]
@H3BO3 No. Like I said in my previous message, it would cause inconsistency. There is "Lerp" and "Lerp Unclamped" for example. Generally the default variant is clamped and if you specifically want unclamped variant there's one specific for that. Lots of things the API follow this principle.

It would be trading off temporary consistency for long term one. If you're a new developer and the nodes by default clamp and have "Unclamped" variant, the expectation would be that "Remap" would follow the same principle.

**Comment by H3BO3** [comment]
Hm, I see. Didn't know about the lerp functions

**Comment by ProbablePrime** [comment]
Thanks for the info Froox, it makes sense :)
